impl/spanner/directory/directory.go

```diff
@@ -0,0 +1,221 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package directory reads and writes Key Transparency Directory information.
+package directory
+
+import (
+	"context"
+	"time"
+
+	"cloud.google.com/go/spanner"
+	"github.com/golang/protobuf/proto"
+	"github.com/golang/protobuf/ptypes"
+	"github.com/golang/protobuf/ptypes/any"
+	"github.com/google/keytransparency/core/directory"
+	"github.com/google/trillian/crypto/keyspb"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	ktspanner "github.com/google/keytransparency/impl/spanner"
+	tpb "github.com/google/trillian"
+)
+
+const table = "Directories"
+
+var directoriesCols = []string{"DirectoryID", "Map", "Log", "VRFPublicKey", "VRFPrivateKey", "MinInterval", "MaxInterval", "Deleted"}
+
+var _ directory.Storage = &Table{}
+
+// Table gives access to the directory table.
+type Table struct {
+	db *ktspanner.Database
+}
+
+// New returns a directory.Storage client backed by Spanner.
+func New(db *ktspanner.Database) *Table {
+	return &Table{db: db}
+}
+
+// spannerDirectory represents one row in the Directories table in spanner.
+type spannerDirectory struct {
+	DirectoryID   string
+	Map           []byte
+	Log           []byte
+	VRFPublicKey  []byte
+	VRFPrivateKey []byte
+	MinInterval   int64
+	MaxInterval   int64
+	Deleted       bool
+}
+
+func readRow(row *spanner.Row) (*directory.Directory, error) {
+	var r spannerDirectory
+	if err := row.ToStruct(&r); err != nil {
+		return nil, err
+	}
+
+	vrfPriv, err := unwrapAnyProto(r.VRFPrivateKey)
+	if err != nil {
+		return nil, err
+	}
+
+	var tmap tpb.Tree
+	if err := proto.Unmarshal(r.Map, &tmap); err != nil {
+		return nil, err
+	}
+	var tlog tpb.Tree
+	if err := proto.Unmarshal(r.Log, &tlog); err != nil {
+		return nil, err
+	}
+
+	return &directory.Directory{
+		DirectoryID: r.DirectoryID,
+		Map:         &tmap,
+		Log:         &tlog,
+		VRF:         &keyspb.PublicKey{Der: r.VRFPublicKey},
+		VRFPriv:     vrfPriv,
+		MinInterval: time.Duration(r.MinInterval) * time.Nanosecond,
+		MaxInterval: time.Duration(r.MaxInterval) * time.Nanosecond,
+		Deleted:     r.Deleted,
+	}, nil
+}
+
+// List returns all Directories. showDeleted indicates whether deleted directories should be returned.
+func (t *Table) List(ctx context.Context, showDeleted bool) ([]*directory.Directory, error) {
+	client, err := t.db.Get(ctx)
+	if err != nil {
+		return nil, err
+	}
+	rtx := client.Single()
+	defer rtx.Close()
+
+	ret := []*directory.Directory{}
+	err = rtx.Read(ctx, table, spanner.AllKeys(), directoriesCols).Do(
+		func(r *spanner.Row) error {
+			d, err := readRow(r)
+			if err != nil {
+				return err
+			}
+			if d.Deleted && !showDeleted {
```

Comment:
Have you considered using SQL to filter those out on the server side? I realise it might be a bit harder code, so in case it looks a good option, add a TODO?

Reply:
done